File: src/xianyu_cli/commands/personal.py
"""Personal commands: profile, favorites, browsing history."""

from __future__ import annotations

import asyncio

import click

from xianyu_cli.core.api_client import TokenExpiredError
from xianyu_cli.core.session import run_api_call
from xianyu_cli.models.envelope import fail, ok
from xianyu_cli.models.user import parse_profile
from xianyu_cli.utils._common import console
from xianyu_cli.utils.credential import load_credential
from xianyu_cli.utils.display import print_items_table, print_profile


def _require_login(output: str):
    cred = load_credential()
    if not cred or cred.is_expired():
        fail("未登录或凭证已过期，请先执行 xianyu login").emit(output)
        return None
    return cred


@click.command()
@click.argument("user_id", required=False, default=None)
@click.pass_context
def profile(ctx: click.Context, user_id: str | None) -> None:
    """查看个人资料（或指定用户的资料）

    示例: xianyu profile
    示例: xianyu profile 12345678
    """
    output = ctx.obj.get("output", "rich")
    cred = _require_login(output)
    if not cred:
        return

    target_id = user_id or cred.user_id

    try:
        result = asyncio.run(
            run_api_call(cred, "mtop.taobao.idle.user.profile", {"userId": target_id})
        )
    except TokenExpiredError:
        fail("登录凭证已过期，请重新执行 xianyu login").emit(output)
        return
    except Exception as e:
        fail(f"获取资料失败: {e}").emit(output)
        return

    parsed = parse_profile(result)

    if output == "rich":
        print_profile(parsed)
    else:
        ok(parsed).emit(output)


@click.command()
@click.pass_context
def favorites(ctx: click.Context) -> None:
    """查看收藏列表"""
    output = ctx.obj.get("output", "rich")
    cred = _require_login(output)
    if not cred:
        return

    try:
        result = asyncio.run(
            run_api_call(cred, "mtop.taobao.idle.user.favorites.list", {"pageSize": 50, "pageNumber": 1})
        )
    except TokenExpiredError:
        fail("登录凭证已过期，请重新执行 xianyu login").emit(output)
        return
    except Exception as e:
        fail(f"获取收藏列表失败: {e}").emit(output)
        return

    items = result.get("itemList", [])

    if output == "rich":
        if items:
            parsed = [
                {
                    "id": i.get("itemId", ""),
                    "title": i.get("title", ""),
                    "price": i.get("price", ""),
                    "location": i.get("area", ""),
                    "seller_name": i.get("userName", ""),
                }
                for i in items
            ]
            print_items_table(parsed, title="收藏列表")
        else:
            console.print("[yellow]收藏列表为空[/yellow]")
    else:
        ok({"items": items}).emit(output)


@click.command()
@click.pass_context
def history(ctx: click.Context) -> None:
    """查看浏览历史"""
    output = ctx.obj.get("output", "rich")
    cred = _require_login(output)
    if not cred:
        return

    try:
        result = asyncio.run(
            run_api_call(cred, "mtop.taobao.idle.user.browsing.history", {"pageSize": 50, "pageNumber": 1})
        )
    except TokenExpiredError:
        fail("登录凭证已过期，请重新执行 xianyu login").emit(output)
        return
    except Exception as e:
        fail(f"获取浏览历史失败: {e}").emit(output)
        return

    items = result.get("itemList", [])

    if output == "rich":
        if items:
            parsed = [
                {
                    "id": i.get("itemId", ""),
                    "title": i.get("title", ""),
                    "price": i.get("price", ""),
                    "location": i.get("area", ""),
                    "seller_name": i.get("userName", ""),
                }
                for i in items
            ]
            print_items_table(parsed, title="浏览历史")
        else:
            console.print("[yellow]暂无浏览历史[/yellow]")
    else:
        ok({"items": items}).emit(output)

File: golang/cmd/personal.go
package cmd

import (
	"fmt"
	"os"

	"github.com/spf13/cobra"

	"xianyu-cli/core"
	"xianyu-cli/models"
	"xianyu-cli/utils"
)

var profileCmd = &cobra.Command{
	Use:   "profile [user_id]",
	Short: "查看个人资料（或指定用户的资料）",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		cred := requireLogin()
		if cred == nil {
			return
		}
		targetID := cred.UserID
		if len(args) > 0 {
			targetID = args[0]
		}

		result, err := core.RunAPICall(cred, "mtop.taobao.idle.user.profile", map[string]interface{}{"userId": targetID})
		if err != nil {
			handleAPIError(err)
			return
		}
		parsed := models.ParseProfile(result)
		if outputMode == "rich" {
			utils.PrintProfile(parsed)
		} else {
			models.OK(parsed).Emit(outputMode)
		}
	},
}

var favoritesCmd = &cobra.Command{
	Use:   "favorites",
	Short: "查看收藏列表",
	Run: func(cmd *cobra.Command, args []string) {
		cred := requireLogin()
		if cred == nil {
			return
		}
		result, err := core.RunAPICall(cred, "mtop.taobao.idle.user.favorites.list", map[string]interface{}{"pageSize": 50, "pageNumber": 1})
		if err != nil {
			handleAPIError(err)
			return
		}
		items, _ := result["itemList"].([]interface{})
		if outputMode == "rich" {
			if len(items) > 0 {
				var parsed []map[string]interface{}
				for _, i := range items {
					m, _ := i.(map[string]interface{})
					parsed = append(parsed, map[string]interface{}{
						"id": m["itemId"], "title": m["title"], "price": m["price"],
						"location": m["area"], "seller_name": m["userName"],
					})
				}
				utils.PrintItemsTable(parsed, "收藏列表")
			} else {
				fmt.Fprintln(os.Stderr, utils.Yellow.Sprint("收藏列表为空"))
			}
		} else {
			models.OK(map[string]interface{}{"items": items}).Emit(outputMode)
		}
	},
}

var historyCmd = &cobra.Command{
	Use:   "history",
	Short: "查看浏览历史",
	Run: func(cmd *cobra.Command, args []string) {
		cred := requireLogin()
		if cred == nil {
			return
		}
		result, err := core.RunAPICall(cred, "mtop.taobao.idle.user.browsing.history", map[string]interface{}{"pageSize": 50, "pageNumber": 1})
		if err != nil {
			handleAPIError(err)
			return
		}
		items, _ := result["itemList"].([]interface{})
		if outputMode == "rich" {
			if len(items) > 0 {
				var parsed []map[string]interface{}
				for _, i := range items {
					m, _ := i.(map[string]interface{})
					parsed = append(parsed, map[string]interface{}{
						"id": m["itemId"], "title": m["title"], "price": m["price"],
						"location": m["area"], "seller_name": m["userName"],
					})
				}
				utils.PrintItemsTable(parsed, "浏览历史")
			} else {
				fmt.Fprintln(os.Stderr, utils.Yellow.Sprint("暂无浏览历史"))
			}
		} else {
			models.OK(map[string]interface{}{"items": items}).Emit(outputMode)
		}
	},
}
